src/keywrap/keyprovider.rs

```diff
@@ -0,0 +1,530 @@
+// Copyright The ocicrypt Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{anyhow, Result};
+use crate::keywrap::KeyWrapper;
+use crate::utils::keyprovider as keyproviderpb;
+use tonic;
+use std::collections::hash_map::RandomState;
+use crate::config::{DecryptConfig, EncryptConfig, KeyProviderAttrs, Command};
+use std::collections::HashMap;
+use crate::{utils};
+use tokio;
+use crate::utils::{CommandExecuter};
+use tokio::runtime::Runtime;
+use core::option::Option;
+use core::fmt::Debug;
+use tonic::codegen::http::Uri;
+
+
+impl Debug for dyn CommandExecuter {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "CommandExecuter")
+    }
+}
+
+/// A KeyProvider keywrapper
+#[derive(Debug)]
+pub struct KeyProviderKeyWrapper {
+    pub provider: String,
+    pub attrs: KeyProviderAttrs,
+    pub runner: Option<Box<dyn CommandExecuter>>,
+}
+
+pub const OP_KEY_WRAP: &str = "keywrap";
+pub const OP_KEY_UNWRAP: &str = "keyunwrap";
+
+/// KeyProviderKeyWrapProtocolInput defines the input to the key provider binary or grpc method.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct KeyProviderKeyWrapProtocolInput {
+    /// op is either "keywrap" or "keyunwrap"
+    #[serde(rename = "op")]
+    op: String,
+    /// keywrapparams encodes the arguments to key wrap if operation is set to wrap
+    #[serde(rename = "keywrapparams")]
+    key_wrap_params: Option<KeyWrapParams>,
+    /// keyunwrapparams encodes the arguments to key unwrap if operation is set to unwrap
+    #[serde(rename = "keyunwrapparams")]
+    key_unwrap_params: Option<KeyUnwrapParams>,
+}
```

Comment:
golang version key_wrap_params/key_unwrap_params did not under Option:
type KeyProviderKeyWrapProtocolInput struct {
        Operation KeyProviderKeyWrapProtocolOperation `json:"op"`
        KeyWrapParams KeyWrapParams `json:"keywrapparams,omitempty"`
        KeyUnwrapParams KeyUnwrapParams `json:"keyunwrapparams,omitempty"`
}

Reply:
key_wrap_params/key_unwrap_params are optional fields, omitempty in json serves the purpose.

Reply:
In golang, struct do not have empty value unless a struct pointer. Then for empty value golang serialized data will be:
```
"DecryptConfig":{"Parameters":{}}
```
But current rust version with Option will be:
```
"DecryptConfig:null"
```
This may make them incompatiable with each other.

Reply:
Yes resolved for keywrapparams/keyunwrapparams and keywrapresults/keyunwrapresults., DecryptConfig in CryptoConfig is a pointer, so left as is for that. 
```
type CryptoConfig struct {
	EncryptConfig *EncryptConfig
	DecryptConfig *DecryptConfig
}
```